### PR TITLE
[WIP] Remote plugins: Add support for perl hosts

### PR DESCRIPTION
**Pull request by jacquesg**
This adds support for creating remote plugins in perl. I've upload the first release of ```Neovim::Ext``` to [CPAN](https://metacpan.org/release/JACQUESG/Neovim-Ext-0.01)

### CODE REVIEW

**Comment by janlazo** [comment]
Needs a couple of steps before this can be merged:
- simple test for `has('perl')`
- include https://github.com/vim/vim/blob/41c363a3154dd1caeb431fa54748ad48ef6d3bd1/src/testdir/test_perl.vim or an earlier version of test_perl.vim to verify that it passes a subset of Vim's test suite
- update `ci/install.sh` and related scripts to install the perl module (cpanm?)
- update `runtime/doc/eval.txt`, `runtime/doc/provider.txt` and related runtime files` to mention `+perl` feature and how to install and use

File: runtime/autoload/provider/perl.vim [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+function! provider#perl#Detect() abort
+  let prog = ''
+  if exists('g:perl_host_prog')
+    let prog = expand(g:perl_host_prog)
+  endif
+
+  " check if perl is on the path
+  if prog == '' && executable('perl')
+    let prog = 'perl'
+  endif
```
Comment (janlazo) [inline runtime/autoload/provider/perl.vim]:
```vim
let prog = get(g:, 'perl_host_prog', executable('perl') ? 'perl' : '')
```

**Comment by jacquesg** [comment]
> * simple test for `has('perl')`
> 
> * include https://github.com/vim/vim/blob/41c363a3154dd1caeb431fa54748ad48ef6d3bd1/src/testdir/test_perl.vim or an earlier version of test_perl.vim to verify that it passes a subset of Vim's test suite

This doesn't add a perl interpreter, but the ability to run plugins externally in perl?

File: runtime/autoload/provider/perl.vim (hunk 1 above)
Comment (jacquesg) [reply runtime/autoload/provider/perl.vim]: Fixed.

**Comment by janlazo** [comment]
> > * simple test for `has('perl')`
> > * include https://github.com/vim/vim/blob/41c363a3154dd1caeb431fa54748ad48ef6d3bd1/src/testdir/test_perl.vim or an earlier version of test_perl.vim to verify that it passes a subset of Vim's test suite
> 
> This doesn't add a perl interpreter, but the ability to run plugins externally in perl?

Vim's `:h feature-list` includes perl. Adding this feature should ideally make `has('perl')` return true if the perl provider is installed correctly. `:h feature-list` includes python, python3, and ruby so these providers enable `:python`, `:python3`, `:ruby` ex commands.

You can do it in a separate PR if it's too hard.

**Comment by justinmk** [comment]
Thanks! Are you aware of https://metacpan.org/pod/Neovim::RPC::API also, and how does this differ? 



> This doesn't add a perl interpreter, but the ability to run plugins externally in perl?

The changes here aren't strictly needed for that though. Typically we would make changes as in this PR to add support for `:perl`, `:perldo`, etc. as @janlazo implied.

The change to `ci/before_install.sh` doesn't do much without some basic test coverage. Besides everything @janlazo said, see also `test/functional/provider/python3_spec.lua` for example of how we test the python interface.

**Comment by jacquesg** [comment]
> Thanks! Are you aware of https://metacpan.org/pod/Neovim::RPC::API also, and how does this differ?

I'm aware of it, it doesn't have support for remote plugins, my first idea was to extend it to add support. I decided against it given the lack of tests in the module and my dislike for its high dependency count. Instead I looked at pynvim for inspiration.

I'll have a look at adding support for ```:perldo``` etc. and creating the basic tests. Obviously the perl module itself has an extensive [testsuite](https://coveralls.io/github/jacquesg/p5-Neovim-Ext).

**Comment by jacquesg** [comment]
Looking at ```:h feature-list``` it seems like ```has('perl')``` is required if the legacy interface is supported? (This PR aims to only implement the remote plugin API)

File: ci/before_install.sh [hunk 2]
```diff
@@ -58,3 +58,7 @@ if [[ -n "$CMAKE_URL" ]]; then
     exit 1
   }
 fi
+
+echo "Install perl"
+curl -L https://install.perlbrew.pl | bash
+perlbrew install --notest perl-5.30.1
```
Comment (janlazo) [inline ci/before_install.sh]: Can you test on the perl version used in Travis? Update `.travis.yml` for apt packages if needed.
Do the same for Appveyor.  Use https://github.com/lukesampson/scoop to install perl and cpanm if Appeveyor has no convenient helpers for perl.

If your plugin must work on the latest perl version only, update your PR description and add a health check to require this latest perl version.

Comment (jacquesg) [reply ci/before_install.sh]: The module does not require a specific version of perl, so any version is fine, I'll update the PR accordingly. As far as I know, supporting remote plugins on Windows is problematic given the issues with ```poll```-ing stdin? To date perl doesn't have an event loop implementation that can actually poll stdin (I've started working on a libuv-based implementation, but that is nowhere near being usable nor complete). I would expect the node provider to work, but I'm unaware of the state of python/ruby providers?

Comment (janlazo) [reply ci/before_install.sh]: https://travis-ci.org/neovim/neovim/jobs/629215538#L1

```
perlbrew version
/home/travis/perl5/perlbrew/bin/perlbrew  - App::perlbrew/0.86
```

Is that recent enough if you need to use perlbrew?

**Comment by janlazo** [comment]
@jacquesg Does your module work with system perl, bundled with OSX, or does it need the homebrew package and C/C++ packages?

**Comment by jacquesg** [comment]
It works with any version of perl.

**Comment by justinmk** [comment]
> Looking at `:h feature-list` it seems like `has('perl')` is required if the legacy interface is supported? (This PR aims to only implement the remote plugin API)

Yeah, it's not a blocker, but generally `:perl`, `:perldo`, etc. are rather easy to implement once the remote API is in place.

**Comment by justinmk** [comment]
Looks like https://github.com/neovim/neovim/pull/11703 will cover this. Thanks @jacquesg !

**Pull request closed without merging**